File: src/controllers/transaction.ts
import { MiddlewareHandler } from "hyper-express";
import { depositETH, depositMNT } from "./../helper/deposit_helper.js";
import { CustomError } from "./../utils/errors.js";
import { withdrawETH, withdrawMNT } from "../helper/withdraw_helper.js";

export const deposit: MiddlewareHandler = async (req, res) => {
  const request = req.body;
  try {
    let totalExecutionTime, txnHash, l1Balance, l2Balance;
    if (request.token === "ETH") {
      const response = await depositETH(request.private_key, request.amount);
      totalExecutionTime = response.totalExecutionTime;
      txnHash = response.txnHash;
      l1Balance = response.ethBalanceOnL1;
      l2Balance = response.ethBalanceOnL2;
    } else if (request.token === "MNT") {
      const response = await depositMNT(request.private_key, request.amount);
      totalExecutionTime = response.totalExecutionTime;
      txnHash = response.txnHash;
      l1Balance = response.mntBalanceOnL1;
      l2Balance = response.mntBalanceOnL2;
    }
    res.json({
      TxnHash: txnHash,
      TotalTimeToTransfer: totalExecutionTime,
      AvailableBalanceOnL1: l1Balance,
      AvailableBalanceOnL2: l2Balance,
    });
  } catch (e) {
    const err = e as CustomError;
    res.json({
      message: err.message,
      error_code: err.code,
    });
  }
};

export const withdraw: MiddlewareHandler = async (req, res) => {
  const request = req.body;
  try {
    let totalExecutionTime, txnHash;
    if (request.token == "ETH") {
      const response = await withdrawETH(request.private_key, request.amount);
      totalExecutionTime = response.totalExecutionTime;
      txnHash = response.txnHash;
    } else if (request.token === "MNT") {
      const response = await withdrawMNT(request.private_key, request.amount);
      totalExecutionTime = response.totalExecutionTime;
      txnHash = response.txnHash;
    }

    res.json({
      TxnHash: txnHash,
      TotalTimeToTransfer: totalExecutionTime,
      Message:
        "Withdraw has initiated from layer 2 to layer 1 chain and it may take several hours before reaching terminal state",
    });
  } catch (e) {
    const err = e as CustomError;
    res.json({
      message: err.message,
      error_code: err.code,
    });
  }
};

File: src/internal/controllers/transaction.go
package controllers

import (
	"net/http"

	"github.com/AbhinitKumarRai/rebalance-funds/src/internal/helper"
	"github.com/AbhinitKumarRai/rebalance-funds/src/pkg/utils"
	"github.com/gin-gonic/gin"
)

// DepositRequest represents the request body for deposit operations
type DepositRequest struct {
	PrivateKey string  `json:"private_key" binding:"required"`
	Amount     float64 `json:"amount" binding:"required"`
	Token      string  `json:"token" binding:"required"`
}

// WithdrawRequest represents the request body for withdraw operations
type WithdrawRequest struct {
	PrivateKey string  `json:"private_key" binding:"required"`
	Amount     float64 `json:"amount" binding:"required"`
	Token      string  `json:"token" binding:"required"`
}

// DepositResponse represents the response for deposit operations
type DepositResponse struct {
	TxnHash              string  `json:"TxnHash"`
	TotalTimeToTransfer  float64 `json:"TotalTimeToTransfer"`
	AvailableBalanceOnL1 float64 `json:"AvailableBalanceOnL1"`
	AvailableBalanceOnL2 float64 `json:"AvailableBalanceOnL2"`
}

// WithdrawResponse represents the response for withdraw operations
type WithdrawResponse struct {
	TxnHash             string  `json:"TxnHash"`
	TotalTimeToTransfer float64 `json:"TotalTimeToTransfer"`
	Message             string  `json:"Message"`
}

// Deposit handles deposit operations
func Deposit(c *gin.Context) {
	var req DepositRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	var response *helper.DepositResponse
	var err error

	switch req.Token {
	case "ETH":
		response, err = helper.DepositETH(req.PrivateKey, req.Amount)
	case "MNT":
		response, err = helper.DepositMNT(req.PrivateKey, req.Amount)
	default:
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid token type"})
		return
	}

	if err != nil {
		if customErr, ok := err.(*utils.CustomError); ok {
			c.JSON(http.StatusInternalServerError, gin.H{
				"message":    customErr.Message,
				"error_code": customErr.Code,
			})
		} else {
			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		}
		return
	}

	c.JSON(http.StatusOK, response)
}

// Withdraw handles withdraw operations
func Withdraw(c *gin.Context) {
	var req WithdrawRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	var response *helper.WithdrawResponse
	var err error

	switch req.Token {
	case "ETH":
		response, err = helper.WithdrawETH(req.PrivateKey, req.Amount)
	case "MNT":
		response, err = helper.WithdrawMNT(req.PrivateKey, req.Amount)
	default:
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid token type"})
		return
	}

	if err != nil {
		if customErr, ok := err.(*utils.CustomError); ok {
			c.JSON(http.StatusInternalServerError, gin.H{
				"message":    customErr.Message,
				"error_code": customErr.Code,
			})
		} else {
			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		}
		return
	}

	c.JSON(http.StatusOK, response)
}
